presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java

```diff
@@ -386,6 +438,239 @@ else if (exchange.getType() == ExchangeNode.Type.REPARTITION) {
+                PartitioningMetadata expectedPartitioningMetadata)
+        {
+            if (!constantExpressions.isEmpty()) {
+                List<Symbol> constantSymbols = ImmutableList.copyOf(constantExpressions.keySet());
+
+                // update outputs
+                outputs = ImmutableList.<Symbol>builder()
+                        .addAll(outputs)
+                        .addAll(constantSymbols)
+                        .build();
+
+                // update inputs
+                inputs = inputs.stream()
+                        .map(input -> ImmutableList.<Symbol>builder()
+                                .addAll(input)
+                                .addAll(constantSymbols)
+                                .build())
+                        .collect(toImmutableList());
+
+                // update sources
+                sources = sources.stream()
+                        .map(source -> {
+                            Assignments.Builder assignments = Assignments.builder();
+                            assignments.putIdentities(source.getOutputSymbols());
+                            constantSymbols.forEach(symbol -> assignments.put(symbol, constantExpressions.get(symbol)));
+                            return new ProjectNode(idAllocator.getNextId(), source, assignments.build());
+                        })
+                        .collect(toImmutableList());
+            }
+
+            NewTableLayout insertLayout = metadata.getInsertLayout(session, tableHandle)
+                    // TODO: support insert into non partitioned table
+                    .orElseThrow(() -> new IllegalArgumentException("insertLayout for the temporary table must be present"));
+
+            PartitioningHandle partitioningHandle = expectedPartitioningMetadata.getPartitioningHandle();
+            List<String> partitionColumns = expectedPartitioningMetadata.getPartitionColumns();
+            ConnectorNewTableLayout expectedNewTableLayout = new ConnectorNewTableLayout(partitioningHandle.getConnectorHandle(), partitionColumns);
+            verify(insertLayout.getLayout().equals(expectedNewTableLayout), "unexpected new table layout");
+
+            Map<String, Symbol> columnNameToSymbol = temporaryTableColumns.entrySet().stream()
+                    .collect(toImmutableMap(entry -> entry.getValue().getName(), Map.Entry::getKey));
+            List<Symbol> partitioningSymbols = partitionColumns.stream()
+                    .map(columnNameToSymbol::get)
+                    .collect(toImmutableList());
+
+            InsertTableHandle insertTableHandle = metadata.beginInsert(session, tableHandle);
+            List<String> outputColumnNames = outputs.stream()
+                    .map(temporaryTableColumns::get)
+                    .map(ColumnMetadata::getName)
+                    .collect(toImmutableList());
+
+            InsertHandle insertHandle = new InsertHandle(insertTableHandle, new SchemaTableName("temporary", "table"));
```

Comment:
I don't understand the constants `"temporary"` `"table"` above.  Perhaps it deserves a comment.  If it's just unused maybe let's create a constant `UNUSED_SCHEMATABLENAME`?

Reply:
Actually i think it is worth setting to the real temporary table schema and name, so it is easier to interpret the plan, and find temporary tables used by a query.

Reply:
Now I'm using `getTableMetadata` to get the real name of the temporary table.